prune/prune.go

```diff
@@ -16,184 +16,209 @@ package prune
+// Pruner is an object that runs a prune job.
+type Pruner struct {
+	registry Registry
 
-	err := config.validate()
-	if err != nil {
-		return err
-	}
+	// client is the controller-runtime client that will be used
+	// To perform a dry run, use the controller-runtime DryRunClient
+	client client.Client
 
-	for i := 0; i < len(config.Resources); i++ {
-		var resourceList []ResourceInfo
-		var err error
-
-		if config.Resources[i].Kind == PodKind {
-			resourceList, err = config.getSucceededPods(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("pods ", "count", len(resourceList))
-		} else if config.Resources[i].Kind == JobKind {
-			resourceList, err = config.getCompletedJobs(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("jobs ", "count", len(resourceList))
-		}
+	// gvk is the type of objects to prune.
+	// It defaults to Pod
+	gvk schema.GroupVersionKind
 
-		var resourcesToRemove []ResourceInfo
-
-		switch config.Strategy.Mode {
-		case MaxAgeStrategy:
-			resourcesToRemove, err = pruneByMaxAge(ctx, config, resourceList)
-		case MaxCountStrategy:
-			resourcesToRemove, err = pruneByMaxCount(ctx, config, resourceList)
-		case CustomStrategy:
-			resourcesToRemove, err = config.CustomStrategy(ctx, config, resourceList)
-		default:
-			return fmt.Errorf("unknown strategy")
-		}
-		if err != nil {
-			return err
-		}
+	// strategy is the function used to determine a list of resources that are pruneable
+	strategy StrategyFunc
 
-		err = config.removeResources(ctx, resourcesToRemove)
-		if err != nil {
-			return err
-		}
-	}
+	// labels is a map of the labels to use for label matching when looking for resources
+	labels map[string]string
 
-	log.V(1).Info("Prune completed")
+	// namespace is the namespace to use when looking for resources
+	namespace string
 
-	return nil
+	// logger is the logger to use when running pruning functionality
+	logger logr.Logger
 }
 
-// containsString checks if a string is present in a slice
-func containsString(s []string, str string) bool {
-	for _, v := range s {
-		if v == str {
-			return true
-		}
+// Unprunable indicates that it is not allowed to prune a specific object.
+type Unprunable struct {
+	Obj    *client.Object
+	Reason string
+}
+
+// Error returns a string representation of an `Unprunable` error.
+func (e *Unprunable) Error() string {
+	return fmt.Sprintf("unable to prune %s: %s", client.ObjectKeyFromObject(*e.Obj), e.Reason)
+}
+
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
+
+// IsPrunableFunc is a function that checks the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `Unprunable` if it is unsafe, or another error.
+// It should safely assert the object is the expected type, otherwise it might panic.
+type IsPrunableFunc func(obj client.Object, logger logr.Logger) error
```

Comment:
We should not include the logger in the signature here. What we have now couples the API to the `logr` package, which is something we very much don't want since there are many go logging options, plus it is likely that one day [go will have its own builtin structured logging interface](https://github.com/golang/go/issues/13182). Users who want a logger in their `IsPruneableFunc` should accomplish this with a closure:

```go
func MyLoggingIsPruneable(logger logr.Logger) IsPruneableFunc {
    return func(obj client.Object) error {
        logr.Info("I love logging!")
    }
}
```